File: problem10.py
s,t = "anagram" ,"nagaram"
# def ValidAnagram():

"""
{"char":"count"}
{"a":2,"n":1,"g":1,r:1,m:1}
"""
def ValidAnagram(s,t):
    set1,set2 = {},{}
    for i in s:
        if i in set1: set1[i]+=1
        else:set1[i] =1
    for j in t:
        if j in set2:set2[j] += 1
        else: set2[j] =1
    return set1==set2
    # print(set1,set2)

print(ValidAnagram(s,t))
print(ValidAnagram("rat","car"))

File: problem10.java
import java.util.HashMap;
import java.util.Map;

public class problem10 {
    // Valid ANAGRAM
    public static boolean validAnagram(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        Map<Character, Integer> map1 = new HashMap<>();
        Map<Character, Integer> map2 = new HashMap<>();
        for (char c : s.toCharArray()) {
            map1.put(c, map1.getOrDefault(c, 0) + 1);
        }
        for (char c : t.toCharArray()) {
            map2.put(c, map2.getOrDefault(c, 0) + 1);
        }
        return map1.equals(map2);
    }

    public static void main(String[] args) {
        System.out.println(validAnagram("anagram", "nagaram"));  // true
        System.out.println(validAnagram("rat", "car"));          // false
        System.out.println();
    }
}

// learn the ascii val of a,A,/, , backspace, enter, null , tab, 0-9
